benchmark/rust-ffi/libacbench/src/lib.rs

```diff
@@ -0,0 +1,44 @@
+extern crate aho_corasick;
+
+use aho_corasick::Automaton;
+use aho_corasick::{AcAutomaton, Dense};
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct U8Slice {
+    ptr: *const u8,
+    off: isize,
+    len: isize,
+}
+
+impl U8Slice {
+    fn into_slice<'a>(&self) -> &'a [u8] {
+        slice_from_pointer(self.ptr, self.off, self.len)
+    }
+}
+
+// https://doc.rust-lang.org/src/core/slice/raw.rs.html#87
+fn slice_from_pointer<'a, T>(ptr: *const T, off: isize, len: isize) -> &'a [T] {
+    &(unsafe { &*std::ptr::slice_from_raw_parts(ptr, (off + len) as usize) })[off as usize..]
+}
+
+#[no_mangle]
+pub extern "C" fn perform_ac(
+    num_needles: isize,
+    needle_slices_: *const U8Slice,
+    haystack_slice_: *const U8Slice,
+) -> isize {
+    let needle_slices = slice_from_pointer(needle_slices_, 0, num_needles);
+    // TODO: Can we somehow allocate this on the stack?
+    let mut needles: Vec<&[u8]> = Vec::with_capacity(num_needles as usize);
+    for i in 0..num_needles as usize {
+        needles.push(needle_slices[i].into_slice());
+    }
+
+    let haystack = slice_from_pointer(haystack_slice_, 0, 1)[0].into_slice();
+
+    let automaton = AcAutomaton::<_, Dense>::with_transitions(&needles[..]);
```

Comment:
I'd be interesting to measure the sparse one as well, as IIRC that's what the Haskell implementation is modeled after.

Reply:
Sure, should I make that the default case? Or maybe add a Cabal flag here similar to the `haskell` benchmark where you can switch between UTF-16 and UTF-8?

Reply:
Yeah, maybe can just add a bool parameter to this function for switching.

Reply:
Addressed in 94bd186